Sundrous fixes

JohnoKing opened the pull request:
- align.c: Remove null pointer arithmetic.
- man pages: Fix mandoc warnings pertaining to lines exceeding 80 bytes.
  - Note that in the man pages, the use of the word [trie](https://en.wikipedia.org/wiki/Trie) is **not** a typo.
- Increase the `SEARCHSIZE` and `LOOKAHEAD` definitions from 80 to 500. (vide https://github.com/ksh93/ksh/issues/460#issuecomment-1340169469)
- `tminit()`: Fixed a small memory leak ASan was complaining about.
- Fixed a compiler error that broke `SHOPT_KIA` by including shlex.h in defs.c (cherrypicked from thickfold-size_t).
- Added pragmas to ignore some superfluous compiler warnings (cherrypicked from thickfold-size_t).
- Added undefs to fix compiler warnings on illumos (cherrypicked from thickfold-size_t).
- ~Fixed undefined behavior in nvtype by using the standard implementation of `alignof` provided by stdalign.h~ Reverted because the fix for that causes a crash in types.sh.

JohnoKing committed:
Sundrous fixes

- align.c: Remove null pointer arithmetic.
- man pages: Fix mandoc warnings pertaining to lines exceeding 80 bytes.
- Increase the SEARCHSIZE and LOOKAHEAD definitions from 80 to 500.
  (vide https://github.com/ksh93/ksh/issues/460#issuecomment-1340169469)
- tminit(): Fixed a small memory leak ASan was complaining about.
- Fixed undefined behavior in nvtype by using the standard implementation
  of alignof provided by stdalign.h rather than the bundled version in
  nvtype.c, fixing the following UBSan error:
  runtime error: member access within null pointer of type '_Align_'
- Fixed a compiler error that broke SHOPT_KIA by including shlex.h
  in defs.c.
- Added pragmas to ignore some superfluous compiler warnings.
- Added undefs to fix compiler warnings on illumos.

JohnoKing committed:
Revert alignof fix for now

It was causing types.sh to crash, so it's currently not upstreamable.

McDutchie commented:
We can't rely on `stdalign.h` anywaqy, because it was only introduced in C11, whereas we're still working on moving to C99.

Can you test this new version of nvtype.c's `alignof` macro instead? It removes the hypothetical null pointer dereference, so this should make UBSan happy. `offsetof` is in C90, so there's no more need to be compatible with compilers without it.

```diff
diff --git a/src/cmd/ksh93/sh/nvtype.c b/src/cmd/ksh93/sh/nvtype.c
index 27a196f9..5acc8d8d 100644
--- a/src/cmd/ksh93/sh/nvtype.c
+++ b/src/cmd/ksh93/sh/nvtype.c
@@ -115,7 +115,7 @@ typedef struct
 	char		*_dpointer;
 } _Align_;
 
-#define alignof(t)	((char*)&((_Align_*)0)->_d##t-(char*)&((_Align_*)0)->_c##t)
+#define alignof(t)	( offsetof(_Align_, _d##t) - offsetof(_Align_, _c##t) )
 
 static void put_type(Namval_t*, const char*, int, Namfun_t*);
 static Namval_t* create_type(Namval_t*, const char*, int, Namfun_t*);
```

JohnoKing committed:
Add offsetof fix; Amend tests to better detect ASan

- Added patch from https://github.com/ksh93/ksh/pull/862#issuecomment-2922604426
  to fix the _Align_ UBSan error.
- Amended the ASan test to check the binary (in case no ASAN_OPTIONS
  flag is present). Also apply a modified form of that test to another
  regression test incompatible with ASan and profiling.

McDutchie committed:
Merge branch 'dev' into sundrous-fixes

McDutchie committed:
add NEWS entry for .sh.edchar length limit increase
